### PR TITLE
feat: refactor monitors

### PR DESCRIPTION
**Pull request by mikasenghaas**
## Summary

Refactor the monitors into a flat `prime_rl.monitors` module (out of `prime_rl.utils.monitor`) with a much smaller surface:

- **One polymorphic entry point.** `monitors.log(...)` (and `Monitor.log(...)`) accepts either a dict of scalar metrics or a list of `vf.Episode` with its cohort coordinates (`kind: train/eval`, `subset: all/effective`); the base class dispatches to `log_metrics` / `log_episodes`. Monitors register once per process via `monitors.setup(...)` into the `MONITORS` list; fan-out catches per-monitor exceptions, so a monitoring failure never takes down training. `NoOpMonitor` and `MultiMonitor` are gone; rank gating lives in `setup`.
- **Side-effect-free construction.** `__init__` takes only the monitor's config (stored by the base class); runtime context is passed to `init()`, whose kwargs each monitor names itself. A monitor whose `init()` raises (e.g. missing API key) crashes the run — a configured monitor must work.
- **Episodes flow through the monitors.** The orchestrator groups rollouts into `vf.Episode`s once (`group_episodes`) and logs each cohort through the same fan-out: the streamed `all` cohort per arrival, the trained `effective` batch per ship, and eval epochs. `FileMonitor` absorbs the trace-file writing (`rollouts/step_N/{train,eval}/{all,effective}/traces.jsonl`, `save_rollouts`/`get_trace_path` inlined); `PrimeMonitor` filters to the trained cohort and uploads one platform sample per episode via verifiers' `build_samples` — the same logic as eval `--push` — through the presigned-parquet flow every 10th step, unsampled (main's previous defaults). `[monitors.prime]` is also available at the `rl` entrypoint's shared level, propagated to the orchestrator only.
- **`TrainRun` owns the platform HTTP lifecycle** (`create`/`log_metrics`/`upload_samples`/`finalize`/`set_status`) with one persistent httpx client (transport-level connect retries) and explicit create fields — the seam the train SDK can subsume later, the way `wandb.Run` backs the W&B monitor. The monitor offloads each call plus the parquet encode via `asyncio.to_thread`, so nothing stalls the event loop; unfinalized runs are marked failed from an `atexit` hook.
- **`$PRL_RUN_ID` is the run identity.** Rollout provenance is stamped from the launcher env (or a local uuid); `Monitor.run_id` and `monitors.run_id()` are gone — the platform run id stays internal to `TrainRun`, and the `$RUN_ID` handshake attaches to a pre-created platform run for managed launches.
- **W&B monitor slimmed** to metric logging and the overview saved view, split into the `monitors/wandb/{monitor,overview}` subpackage.

## Breaking

- **Config layout**: monitor configs are grouped under `[monitors]` — `[monitors.wandb]`, `[monitors.file]`, and orchestrator-only `[orchestrator.monitors.prime]` replace `[wandb]`, `[file_monitor]`, and `[orchestrator.prime_monitor]`. Config classes live in `prime_rl.configs.monitors`; `WandbConfig` is renamed `WandbMonitorConfig`, `FileMonitorConfig.filename` becomes `path`, and `PrimeMonitorConfig` shrinks to just `name` (renamed from `run_name`, still auto-set from `run.name` / the W&B name); `team_id` comes from the Prime CLI config — base URL and API key env var are hardcoded, the frontend URL comes from the Prime CLI config.
- **The file monitor is on by default** (`--no-monitors.file` to opt out); W&B stays opt-in. Trace files are written by the file monitor now, so disabling it also disables the `rollouts/` trace streams.
- **Distribution logging removed**: `log_extras.distributions` and the `/distributions` upload are gone; `log_extras` has no replacement — episodes upload every 10th step, unsampled (main's previous defaults), to be revisited with the prime traces SDK.
- **W&B sample tables removed**: sample inspection lives in the platform episode viewer and the trace files.
- **Run identity**: `$PRL_RUN_ID` is the authoritative identity stamped on rollouts (`Monitor.run_id` / `monitors.run_id()` are gone; the platform run id stays internal to `TrainRun`). The `$RUN_ID` attach handshake is kept: managed launches that inject it attach to the pre-created platform run instead of registering a duplicate — backends should inject `PRL_RUN_ID=<platform run id>` alongside it so W&B and trace provenance align. Companion: [fft#56](https://github.com/PrimeIntellect-ai/fft/pull/56).
- **Usage reporter removed**: the `PI_USAGE_*`-gated per-step training-token billing posts are gone, along with `prime_rl/utils/usage_reporter.py` and `background_async.py`.
- **Python API**: `prime_rl.utils.monitor` (and `setup_monitor`/`get_monitor`) is replaced by `prime_rl.monitors`; `save_final_summary` becomes `finalize` (platform-run finalization only — the local `final_summary.json` files had no consumers and are no longer written).

## Verification

Final e2e on reverse-text with all three monitors (wandb + file on by default, prime enabled):

- **Happy path** (10 steps): zero monitor warnings; one init + one finalize info log per monitor; episodes uploaded every step; `rollouts/step_{1..10}/train/{all,effective}/traces.jsonl` written by the file monitor; `metrics.jsonl` with steps 1-10 from both processes; platform run finalized **COMPLETED** ([dashboard](https://app.primeintellect.ai/dashboard/training/n0i4hpa85vff0riu9di6p0s9), verified via the API's terminal-state 409).
- **Unhappy path** (injected raise before step 1): `finalize` skipped (`clean_exit` gate), `TrainRun`'s atexit hook marked the platform run **FAILED** ([dashboard](https://app.primeintellect.ai/dashboard/training/lda49ihi46f7ppad0kxkjkf6), verified via 409); W&B decays to *crashed* via heartbeat timeout.
- Known upstream gap: W&B run *state* in (experimental) shared mode never reflects `finish` — even clean runs with an explicit `wandb.finish(exit_code=0)` decay to *crashed* after ~10 min. Pre-existing (this morning's pre-refactor run shows the same) and not client-fixable; metrics/history are unaffected.

`tests/unit` passes; functional smokes cover the polymorphic `log` overloads (dict / episode / episodes), trace-file cohorts, the prime cohort filter, crash-on-init, and the atexit lifecycle.

---

> [!NOTE]
> **Medium Risk**
> Large, breaking observability refactor (config keys, default file monitor, removed W&B extras and usage reporting) across orchestrator, trainer, and SFT entrypoints; core training logic is largely unchanged but misconfigured runs may silently lose traces or fail platform init.
> 
> **Overview**
> Monitoring moves from `prime_rl.utils.monitor` (`setup_monitor`, `MultiMonitor`/`NoOpMonitor`) to **`prime_rl.monitors`**: rank-0 registers W&B, file, and Prime backends via `monitors.setup`, training code calls **`monitors.log`** for scalar dicts or **`vf.Episode`** lists with `kind`/`subset` (`train`/`eval`, `all`/`effective`), and clean exits call **`monitors.finalize`**. Per-monitor failures are caught so logging cannot crash training.
> 
> **Config is a breaking rename**: top-level/shared **`[wandb]`** → **`[monitors.wandb]`**, **`[file_monitor]`** → **`[monitors.file]`** ( **`path`** instead of `filename`, **on by default** — `--no-monitors.file` to disable), and **`[orchestrator.prime_monitor]`** → **`[monitors.prime]`** / **`[orchestrator.monitors.prime]`** with **`name`** only. Shared `rl` propagation and validation in `packages/prime-rl-configs` follow the new paths; example/CI TOMLs and docs are updated accordingly.
> 
> The orchestrator no longer calls `save_rollouts` / `get_trace_path` or separate W&B sample/distribution hooks. It **`group_episodes`** and logs cohorts through monitors: **`all`** on arrival, **`effective`** at train ship / eval finalize. **`FileMonitor`** writes `metrics.jsonl` and `rollouts/step_N/.../traces.jsonl`. **`PrimeMonitor`** uses async **`TrainRun`** (metrics + presigned Parquet episode uploads every 10th step for train/effective). **`WandbMonitor`** is metrics + overview view only — **`log_extras`** sample/distribution tables are removed.
> 
> Also removed: **`UsageReporter`** / `PI_USAGE_*` billing posts, **`save_final_summary`** / local `final_summary.json`, and rollout identity from monitor `run_id` ( **`$PRL_RUN_ID`** for provenance; **`$RUN_ID`** still attaches pre-created platform runs).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c3141f4ab39407dc13489dbc28dc3123f25ebb21. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/monitors/prime.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+    async def log_episodes(self, episodes: list[vf.Episode], step: int, kind: Kind, subset: Subset) -> None:
+        """Upload one platform sample per episode via the presigned-URL Parquet flow.
+        Only the trained cohort ships to the platform."""
+        if kind != "train" or subset != "effective" or not episodes:
+            return
+
+        # Log 10% of the step's episodes (at most 128) to not overwhelm the
+        # platform's ingestion. TODO: Lift once we integrate the prime traces SDK.
+        episodes = episodes[: min(max(1, len(episodes) // 10), 128)]
```
[the remark this answers is not in the document]
Comment (mikasenghaas) [reply src/prime_rl/monitors/prime.py]: Fixed in d8007047a — back to `random.sample` with a comment explaining why a prefix is biased (batch is ordered by group finalization).

File: .gitignore [hunk 2]
```diff
@@ -17,6 +17,7 @@ configs/endpoints.py
 **/rollouts
 **/logs
 **/wandb
+!src/prime_rl/monitors/wandb/
```
Comment (samsja) [inline .gitignore:20]: what is this ? why "!"

Comment (samsja) [reply .gitignore:20]: oh gotcha bc we remove /wandb above

**Review by samsja — APPROVED** [review]

File: .gitignore (hunk 2 above)
Comment (mikasenghaas) [reply .gitignore:20]: ya could prob do nicer but the **/wandb gitignore (which we do to not push wandb outputs) also matched the code file😭 but ya like whatever
